app/js/arethusa.search/services/search.js

```diff
@@ -40,6 +40,182 @@ angular.module('arethusa.search').service('search', [
       });
     };
 
+    this.findWordInContext = function(query) {
+
+      return arethusaUtil.inject([], self.strings, function (memo, string, ids) {
+        var matches = []
+        angular.forEach(ids, function(id) {
+          var previousToken = state.getPreviousTokens(id,1);
+          var nextToken = state.getNextTokens(id,1);
+          var nextString = nextToken.length > 0 ? nextToken[0].string : null;
+          var prevString = previousToken.length > 0 ? previousToken[0].string : null;
+          var match = self.compareWordsWithContext(string,prevString,nextString,query.word);
+          if (match.match) {
+            var matchData = {
+              id: id,
+              matchedPrefix: 0,
+              matchedSuffix: 0
+            }
+            if (match.combine < 0) {
+              matchData.includePrevious = previousToken[0].id;
+            } else if (match.combine > 0) {
+              matchData.includeNext = nextToken[0].id;
+            }
+            matches.push(matchData)
+          }
+        });
+        // we only need to try to further narrow by context if we got more than
+        // one hit
+        if (matches.length > 1) {
+          matches.forEach(function (match) {
+            var startingId = match.includePrefix ? match.includePrefix :
+              match.includeSuffix ? match.includeSuffix : match.id;
+            var prefixWords = tokenize(query.prefix);
+            var suffixWords = tokenize(query.suffix);
+            var matchedPrefixWords = 0;
+            var matchedSuffixWords = 0;
+            var previousTokens = [];
+            if (prefixWords.length > 0) {
+              var previousTokens = state.getPreviousTokens(startingId);
+              var tokenIndex = previousTokens.length-1;
+              for (var i=prefixWords.length-1; i>=0; i--) {
+                if (previousTokens[tokenIndex]) {
+                  var token = previousTokens[tokenIndex];
+                  var nextString = previousTokens[tokenIndex+1] ? previousTokens[tokenIndex+1].string : null;
+                  var prevString = tokenIndex > 0 && previousTokens[tokenIndex-1] ? previousTokens[tokenIndex-1].string : null;
+                  var matchP = self.compareWordsWithContext(token.string,prevString,nextString,prefixWords[i]);
+                  if (matchP.match) {
+                    tokenIndex--;
+                    matchedPrefixWords++;
+                    tokenIndex = tokenIndex + matchP.combine;
```

Comment:
I'm having some trouble following this logic, but if `combine` is `1` then that means that the word at `tokenIndex` `a` combined with the word `tokenIndex` `a + 1` matches the word being tested? In that case don't we still want to decrement `tokenIndex`? Otherwise it ends up testing the exact same word in the next iteration of the loop.

Reply:
ah, yes I think you're right.  we always do decrement token index, but I missed the fact that if combine == 1 that we are undoing that.